aws/auth/mutable_static_creds_provider.cc

```diff
@@ -0,0 +1,142 @@
+#endif
+}
+
+using namespace aws::auth;
+
+MutableStaticCredentialsProvider::MutableStaticCredentialsProvider(const std::string& akid,
+                                                                   const std::string& sk,
+                                                                   std::string token) {
+  Aws::Auth::AWSCredentials creds(akid, sk, token);
+  current_.creds_ = creds;
+}
+
+void MutableStaticCredentialsProvider::set_credentials(const std::string& akid, const std::string& sk, std::string token) {
+  std::lock_guard<std::mutex> lock(update_mutex_);
+
+  //
+  // The ordering stores here is important.  current_.updating_, and current_.version_ are atomics.
+  // Between the two of them they create an interlocked state change that allows consumers
+  // to detect that the credentials have changed during the process of copying the values.
+  //
+  // Specifically current_.version_ must be incremented before current_.updating_ is set to false.
+  // This ensures that consumers will either see a version mismatch or see the updating state change.
+  //
+
+  current_.updating_ = true;
+  current_.creds_.SetAWSAccessKeyId(akid);
+  current_.creds_.SetAWSSecretKey(sk);
+  current_.creds_.SetSessionToken(token);
+  current_.version_++;
+  current_.updating_ = false;
+ }
+
+Aws::Auth::AWSCredentials MutableStaticCredentialsProvider::GetAWSCredentials() {
+
+  Aws::Auth::AWSCredentials result;
+  //
+  // This is an attempt to do an optimistic read.  We assume that the contents of the
+  // credentials may change in while copying to the result.
+  //
+  // 1. To handle this we first check if an update is in progress, and if it is we bounce out
+  // and retry.
+  //
+  // 2. If the credential isn't being updated we go ahead and load the current version of
+  // the credential.
+  //
+  // 3. At this point we go ahead and trigger a copy of the credential to the result.  This
+  // is what we will return if our remaining checks indicate everything is still ok.
+  //
+  // 4. We check again to see if the credential has entered updating, if it has it's possible
+  // the version of the credential we copied was split between the two updates.  So we
+  // discard our copy, and try again.
+  //
+  // 5. Finally we make check to see that the version hasn't changed since we started.  This
+  // ensures that the credential didn't enter and exit updates between our update checks.
+  //
+  // If everything is ok we are safe to return the credential, while it may be a nanosecnds
+  // out date it should be fine.
+  //
+
+  std::uint64_t starting_version = 0, ending_version = 0;
+  do {
+    if (current_.updating_) {
+      //
+      // The credentials are currently being updated.  It's not safe to read so
+      // spin while we wait for the update to complete.
+      //
+      update_step(debug_stats.update_before_load_, debug_stats.retried_);
+      continue;
+    }
+    std::uint64_t starting_version = current_.version_.load();
```

Comment:
should we be consistent in our use of atomic. If we use the `atomic<T>::operator T` then we should use it everywhere instead of the equivalent `.load` method unless we really want to specify memory order.

Reply:
Fixed